### ISSUE TITLE
Multiple critical issues

### ISSUE THREAD

**Issue by lyphyser** [mitsuhiko/self-replace#4]
I looked at the code, and it's not really in a production-usable state.

There are mostly for the Windows version:

1. If fs::copy is interrupted in the middle (e.g. by power failure or killing the process), the new file will be corrupted and the old one will be deleted, resulting in an unrecoverable catastrophe. Instead, fs::copy must write a temporary file, and only afterwards (and after an fsync) be atomically renamed to the executable name, and also deletion of the current executable must happen only after the rename is successful
2. The code apparently only tries a single name for the temporary files, which might well already exists. Code that creates temporary files must of course retry until it finds a name that doesn't exist.
3. If there is a power failure, temporary files are left as-is and there is no recovery. Instead, on Windows, it should use PendingFileRenameOperations to recover, and on Unix probably a bunch of mechanics (e.g. creating a systemd unit if systemd is installed, with perhaps fallback to editing ~/.profile)
4. It should try to open the current executable with FILE_DELETE_ON_CLOSE and if that succeeds there is no need to do the respawn trick
5. There is a sleep for 100ms in the code. Such sleeps are always wrong since an sleeping for an arbitrary amount of time serves no purpose. Instead, either wait for what you need or retry in a loop.
6. Doing the self delete on init by checking the executable suffix is completely broken, since the user might have to decided to name the executable with a name having such a suffix. Instead, the code should be work by dropping a separately compiled executable that does only the deletion instead of copying the current executable, or if that's undesirable, by using an environment variable with a randomly-generated name.
7. As mentioned in the code, the Rust std functions must not be used before main (this is not a problem if a separate executable is used).
8. The code spawns a "ping" process at some point, which is completely absurd (among others, there may be no ping, no network stack, no 127.0.0.1 address, not to mention that libraries should not spawn random processes for no reason, etc.). That should be replaced with a proper solution that does whatever is needed. If you absolutely need to spawn a system executable (which I doubt and hope not) then at least spawn something that is reasonably guaranteed to exist, such as rundll32
9. The code uses file paths for all operations, which means that it's going to fail spectacularly if the user renames a file in the middle of the operation. Instead, file paths must be used once for opening and then fd/handle-based operations (the *at APIs on Unix) must be used afterwards. If that's not possible, there must at least be code to retry the whole thing if a file gets renamed by the user.
10. The code doesn't work at all if the executable has been deleted or renamed. It should detect that case and not do the deletion.
11. The code seems to assume that once the parent shuts down, then the executable is deletable, but this is obviously false, since something else might have opened the executable without FILE_SHARE_DELETE. It should handle this case by either waiting until deletion is possible, or scheduling it for the next reboot.
12. If you really need to get the parent process id, then you should use InheritedFromUniqueProcessId from NtQueryInformationProcess and not toolhelp32 (which is O(n) in the number of processes in the system).
13. The unix code sets file permissions after doing the copy, which is obviously a security disaster since the executable might be readable and even writable by everyone (or by the group) in the meantime if umask is set to maximally permissive

This is just the issues that I have noticed without really doing an exhaustive check. Given the apparent lack of any thought to most of these issues, I would expect more to be present and would not really recommend using this crate, at least until it's pretty much rewritten.

**Comment by mitsuhiko** [comment]
> If fs::copy is interrupted in the middle (e.g. by power failure or killing the process), the new file will be corrupted and the old one will be deleted, resulting in an unrecoverable catastrophe

You are right that on the windows case the atomic write is missing, that has been changed. The `fsync` part you mentioned I don't think is particularly relevant. I have added the missing rename.

> The code apparently only tries a single name for the temporary files

It uses 20 random characters for entropy, it's doubtful that this would ever clash.

> If there is a power failure, temporary files are left as-is and there is no recovery

Yeah, that is a known limitation. I filed #5 for it, but I don't think the issue is particularly critical as the race for this is very tight. The files placed in temp locations particularly on windows are not affected by this anyways as the operating system will clean them up. On Windows you will in fact only lose files if the volume of the target exe and the temp folder is not shared.

> It should try to open the current executable with FILE_DELETE_ON_CLOSE and if that succeeds there is no need to do the respawn trick

There is no known version of Windows where this works, so I would not even know how to test this.

> Doing the self delete on init by checking the executable suffix is completely broken, since the user might have to decided to name the executable with a name having such a suffix.

I don't consider this broken, but it's undocumented behavior and maybe a more random suffix should be used to reduce the likelihood of this happening. I filed #6 to track this.

> As mentioned in the code, the Rust std functions must not be used before main

Yeah, I will eventually get to this and replace them with native win32 code. That said, the section I picked for the function is run after the runtime initialized. Tracking in #7.

> The code spawns a "ping" process at some point, which is completely absurd (among others, there may be no ping, no network stack, no 127.0.0.1 address, not to mention that libraries should not spawn random processes for no reason, etc.). That should be replaced with a proper solution that does whatever is needed. If you absolutely need to spawn a system executable (which I doubt and hope not) then at least spawn something that is reasonably guaranteed to exist, such as rundll32

The original code spawned rundll32. Unfortunately rundll32 can in theory shut down quite quickly. I am not aware of a version of Windows that does not have ping available so that concern does not seem particularly high. For the hypothetical case that there is no 127.0.0.1 the spawn of ping might be slow enough anyways to solve this issue. The spawn of an executable is necessary and the motivation is documented.

> The code uses file paths for all operations, which means that it's going to fail spectacularly if the user renames a file in the middle of the operation.

Since the files in all cases need to appear in the file system there are no alternatives to this. This issue however applies to most uses of filesystem APIs and is not specific to this crate. Any "create file and move over atomically" operation has the same issue.

> The code doesn't work at all if the executable has been deleted or renamed. It should detect that case and not do the deletion.

That is intentional.

> The code seems to assume that once the parent shuts down, then the executable is deletable, but this is obviously false, since something else might have opened the executable without FILE_SHARE_DELETE. It should handle this case by either waiting until deletion is possible, or scheduling it for the next reboot.

This only applies to the copied executable which nobody can launch in the meantime unless they race intentionally for it to break it. There are however cases where the deletion of the main executable after the rename cannot work. This is basically a duplicate of #5 

> If you really need to get the parent process id, then you should use InheritedFromUniqueProcessId from NtQueryInformationProcess and not toolhelp32 (which is O(n) in the number of processes in the system).

I considered this, however the behavior of the current version was battle tested by rustup so I decided to keep it. The issue with `InheritedFromUniqueProcessId` is that in theory this is not the parent process and I did not want to investigate yet if this is an issue. If the performance is a problem I'm sure an issue will be opened. In general that operation is not particularly common anyways.

> The unix code sets file permissions after doing the copy, which is obviously a security disaster since the executable might be readable and even writable by everyone (or by the group) in the meantime if umask is set to maximally permissive

I don't believe this to be an actual problem given that this operation is performed on a temporary. Maybe you could elaborate on the exact way in which this might be exploitable in a new issue.

I opened specific issues for the individual points that that I believe need to be addressed.
